src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
-			if(str_index < str_len && true) { //mem_readb(BIOS_KEYBOARD_FLAGS1)&0x80) dev_con.h ?
-				outc(' ');//move cursor one to the right.
-				auto text_len = static_cast<uint16_t>(str_len - str_index);
-				uint8_t* text=reinterpret_cast<uint8_t*>(&line[str_index]);
-				DOS_WriteFile(STDOUT, text, &text_len); // write buffer to screen
-				move_cursor_back_one(); //undo the cursor the right.
-				for (auto i = str_len; i > str_index; i--) {
-					line[i]=line[i-1]; //move internal buffer
-					move_cursor_back_one(); //move cursor back (from write buffer to screen)
-				}
-				// new end (as the internal buffer moved one
-				// place to the right
-				terminate_str_at(line, ++str_len);
-				size--;
-			};
-
-			line[str_index]=c;
-			str_index ++;
-			if (str_index > str_len){ 
-				terminate_str_at(line, str_index);
-				str_len++;
-				size--;
+			if (Int10Manager::MaxCommandSize() < command.size()) {
+				break;
 			}
-			DOS_WriteFile(STDOUT,&c,&n);
+			command.insert(cursor_position, 1, static_cast<char>(data));
+			cursor_position++;
 			break;
 		}
+
+		int10_manager.Update(command, cursor_position);
+
+		if (!viewing_tab_completions) {
+			completion.clear();
+		}
 	}
 
-	if (!str_len) return;
-	str_len++;
+	return "";
+}
+
+namespace {
+std::vector<std::string> get_completions(const std::string_view command)
+{
+	const std::vector<std::string> args = split(command);
 
-	// remove current command from history if it's there
-	if (current_hist) {
-		// current_hist=false;
-		l_history.pop_front();
+	const bool dir_only = (!args.empty() && iequals(args[0], "CD") &&
+	                       (args.size() > 1 || command.back() == ' '));
+
+	std::string search = {};
+	if (!args.empty() && command.back() != ' ') {
+		search += args.back();
 	}
+	search += "*.*";
+
+	auto save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
 
-	// add command line to history
-	l_history.push_front(line); it_history = l_history.begin();
-	if (l_completion.size()) l_completion.clear();
-
-	/* DOS %variable% substitution */
-	const auto dos_section = static_cast<Section_prop *>(control->GetSection("dos"));
-	assert(dos_section);
-	std::string_view expand_shell_variable_pref =
-					dos_section->Get_string("expand_shell_variable");
-	if (expand_shell_variable_pref == "true") {
-		ProcessCmdLineEnvVarStitution(line);
-	} else if (expand_shell_variable_pref == "auto" && dos.version.major >= 7) {
-		ProcessCmdLineEnvVarStitution(line);
+	DOS_DTA::Result result = {};
+
+	bool res = DOS_FindFirst(search.c_str(), ~DOS_ATTR_VOLUME);
+
+	std::vector<std::string> files           = {};
+	std::vector<std::string> non_executables = {};
+	while (res) {
+		dta.GetResult(result);
+
+		if (!result.IsDummyDirectory() &&
+		    (!dir_only || result.IsDirectory())) {
+			if (is_executable_filename(result.name)) {
+				files.emplace_back(result.name);
+			} else {
+				non_executables.emplace_back(result.name);
```

Comment:
Maybe move the string's in: `.push_back(std::move(result.name))`.

I guess we'd have to also place the `DOS_DTA::Result result = {};` inside the while look prior to getting the result, just to start with a fresh object each time.